daemon/cluster/controllers/plugin/controller.go

```diff
@@ -1,32 +1,149 @@
+// With the the plugin controller instead of having a strict create->start->stop->remove
+// task lifecycle like containers, we manage the desired state of the plugin and let
+// the plugin manager do what it already does and monitor the plugin.
+// We'll also end up with many tasks all pointing to the same plugin ID.
+type Controller struct {
+	backend   Backend
+	spec      swarmtypes.GenericPluginPayload
+	id        string
+	serviceID string
+}
+
+// Backend is the interface for interacting with the plugin manager
+// Controller actions are passed to the configured backend to do the real work.
+type Backend interface {
+	Disable(name string, config *enginetypes.PluginDisableConfig) error
+	Enable(name string, config *enginetypes.PluginEnableConfig) error
+	Remove(name string, config *enginetypes.PluginRmConfig) error
+	Pull(ctx context.Context, ref reference.Named, name string, metaHeaders http.Header, authConfig *enginetypes.AuthConfig, privileges enginetypes.PluginPrivileges, outStream io.Writer, opts ...plugin.CreateOpt) error
+	Upgrade(ctx context.Context, ref reference.Named, name string, metaHeaders http.Header, authConfig *enginetypes.AuthConfig, privileges enginetypes.PluginPrivileges, outStream io.Writer) error
+	Get(name string) (*v2.Plugin, error)
+	SubscribeEvents(buffer int, events ...plugin.Event) (eventCh <-chan interface{}, cancel func())
+}
 
 // NewController returns a new cluster plugin controller
-func NewController() (*Controller, error) {
-	return &Controller{}, nil
+func NewController(backend Backend, t *api.Task) (*Controller, error) {
+	spec, err := readPayload(t)
+	if err != nil {
+		return nil, err
+	}
+	return &Controller{backend: backend, spec: spec, serviceID: t.ServiceID}, nil
 }
 
 // Update is the update phase from swarmkit
 func (p *Controller) Update(ctx context.Context, t *api.Task) error {
 	logrus.WithFields(logrus.Fields{
 		"controller": "plugin",
 	}).Debug("Update")
+
 	return nil
 }
 
+func readPayload(t *api.Task) (swarmtypes.GenericPluginPayload, error) {
+	var cfg swarmtypes.GenericPluginPayload
+
+	generic := t.Spec.GetGeneric()
+	if err := json.Unmarshal(generic.Payload.Value, &cfg); err != nil {
+		return cfg, errors.Wrap(err, "error reading plugin config")
+	}
+	return cfg, nil
+}
+
 // Prepare is the prepare phase from swarmkit
-func (p *Controller) Prepare(ctx context.Context) error {
+func (p *Controller) Prepare(ctx context.Context) (err error) {
 	logrus.WithFields(logrus.Fields{
 		"controller": "plugin",
 	}).Debug("Prepare")
+
+	// TODO(@cpuguy83): support digest pinning
+	remote, err := reference.ParseNormalizedNamed(p.spec.Spec.Remote)
+	if err != nil {
+		return errors.Wrapf(err, "error parsing remote reference %q", p.spec.Spec.Remote)
+	}
+
+	type canonicalWithTag interface {
+		reference.Canonical
+		Tag() string
+	}
+
+	if canonical, ok := remote.(canonicalWithTag); ok {
+		remote, err = reference.WithDigest(reference.TrimNamed(remote), canonical.Digest())
```

Comment:
It should be possible to pass a reference with both a tag and a digest to the pull code. This wasn't working properly earlier, but it was fixed recently.

Not sure if there are any plugin-specific issues with doing this, but ideally there shouldn't be.